### ISSUE TITLE
Python 2 vs 3: 'DenseLayer' object has no attribute 'input_shape'

### ISSUE THREAD

**Issue by urinieto** [Lasagne/Lasagne#297]
I am trying to use a model trained on Python 2.7 (which works fine on Python 2.7) with a Python 3.4 script, but I get the following error:

```
File "/home/uri/Projects/segmenter/segmenter.py", line 31, in get_cnn_output
    batch_size = lasagne.layers.get_output_shape(model)[0]
  File "/home/uri/Projects/Lasagne/lasagne/layers/helper.py", line 266, in get_output_shape
    return layer_or_layers.output_shape
  File "/home/uri/Projects/Lasagne/lasagne/layers/base.py", line 46, in output_shape
    return self.get_output_shape_for(self.input_shape)
nose.proxy.AttributeError: 'DenseLayer' object has no attribute 'input_shape'
```

Is this a bug or this is simply not supported?

Thanks!

**Comment by benanne** [comment]
Are you using the latest version from git? `DenseLayer`most definitely has an attribute `input_shape` after initialization, it's set in the base class constructor: https://github.com/Lasagne/Lasagne/blob/master/lasagne/layers/base.py#L34

So I don't really understand where this error comes from. It's also a `nose.proxy.AttributeError`, which is weird because we do not use nose :) Maybe you use it in your code somewhere? I don't think this is a Python 2/3 issue, at any rate.

**Comment by urinieto** [comment]
Yes, I'm using the latest version from the repo (I just pulled around an hour ago).

Sorry for the confusion, the error I reported occurred within a `nose` unit test. Here you have the same error reported by iPython outside any unit testing framework (again, this works totally fine on Python 2.7):

```
/home/uri/Projects/segmenter/segmenter.py in get_cnn_output(input_patches, model)
     29     """
     30     # Get Batch Size of the network
---> 31     batch_size = lasagne.layers.get_output_shape(model)[0]
     32 
     33     x = input_patches.reshape(-1, 1, config.FRAMES_CONTEXT, config.CQT_BINS)

/home/uri/Projects/Lasagne2/lasagne/layers/helper.py in get_output_shape(layer_or_layers, input_shapes)
    264             return [layer.output_shape for layer in layer_or_layers]
    265         except TypeError:
--> 266             return layer_or_layers.output_shape
    267 
    268     from .input import InputLayer

/home/uri/Projects/Lasagne2/lasagne/layers/base.py in output_shape(self)
     50     @property
     51     def output_shape(self):
---> 52         return self.get_output_shape_for(self.input_shape)
     53 
     54     def get_params(self, **tags):

AttributeError: 'DenseLayer' object has no attribute 'input_shape'
```

**Comment by dnouri** [comment]
> I am trying to use a model trained on Python 2.7 (which works fine on Python 2.7) with a Python 3.4 script, but I get the following error:

You're saying you've trained and pickled a model with 2.7 and loading it back in with Python 3.4?

**Comment by urinieto** [comment]
> You're saying you've trained and pickled a model with 2.7 and loading it back in with Python 3.4?

Exactly. Is this a bad idea / practice?

**Comment by ebenolson** [comment]
Are you sure you're using the same Lasagne version in both 2.7 and 3.4?

**Comment by urinieto** [comment]
> Are you sure you're using the same Lasagne version in both 2.7 and 3.4? 

I'm pretty sure.

**Comment by craffel** [comment]
Hola titan,

> Exactly. Is this a bad idea / practice?

Better practice/more portable would be to save the model parameters using `lasagne.layers.get_all_param_values` and load using `lasagne.layers.set_all_param_values`.

> I'm pretty sure.

You should make sure :)

**Comment by ebenolson** [comment]
I was able to dump and load a simple network `[InputLayer, DenseLayer, DenseLayer]` successfully between 2.7 and 3.4. My guess is your pickle was saved from an older version of Lasagne.

**Comment by urinieto** [comment]
> Better practice/more portable would be to save the model parameters using `lasagne.layers.get_all_param_values` and load using `lasagne.layers.set_all_param_values`.

Nice, I like that. Thanks titaner!

> You should make sure :)

I did :)

> I was able to dump and load a simple network [InputLayer, DenseLayer, DenseLayer] successfully between 2.7 and 3.4. My guess is your pickle was saved from an older version of Lasagne.

I think that could be it. I'm gonna check it right now. I will report back. Thanks!

**Comment by urinieto** [comment]
I figured it out. Turns out it had nothing to do with Lasagne, but with some [pickle incompatibilities] (http://stackoverflow.com/questions/11305790/pickle-incompatability-of-numpy-arrays-between-python-2-and-3) between Python 2 and 3.

I fixed it by "unpickling" the model like this in Python 3.4:

```
with open(model_name, "rb") as f:
    model = pickle.load(f, encoding="latin1")
```

(I was using `encoding="bytes"` before, which produced the error I reported above)

Thanks for your help in any case!
